packages/transaction-controller/src/utils/gas-fees.ts

```diff
@@ -12,33 +12,61 @@ import { GAS_ESTIMATE_TYPES } from '@metamask/gas-fee-controller';
 import { createModuleLogger } from '@metamask/utils';
 import { addHexPrefix } from 'ethereumjs-util';
 
+import { SWAP_TRANSACTION_TYPES } from '../constants';
 import { projectLogger } from '../logger';
-import type { TransactionParams } from '../types';
-import { UserFeeLevel, type TransactionMeta } from '../types';
+import type {
+  AdvancedGasFees,
+  TransactionParams,
+  TransactionMeta,
+  TransactionType,
+} from '../types';
+import { UserFeeLevel } from '../types';
 
 export type UpdateGasFeesRequest = {
   eip1559: boolean;
   ethQuery: EthQuery;
+  getAdvancedGasFee: () => AdvancedGasFees;
   getGasFeeEstimates: () => Promise<GasFeeState>;
+  isAdvancedGasFeeDisabled: boolean;
   txMeta: TransactionMeta;
 };
 
 export type GetGasFeeRequest = UpdateGasFeesRequest & {
+  advancedGasFees?: AdvancedGasFees;
   initialParams: TransactionParams;
+  shouldReadAdvancedGasFees: boolean;
   suggestedGasFees: Awaited<ReturnType<typeof getSuggestedGasFees>>;
 };
 
+/**
+ * Represents the user customizing their gas preference
+ */
+export const CUSTOM_GAS_ESTIMATE = 'custom';
+
 const log = createModuleLogger(projectLogger, 'gas-fees');
 
 export async function updateGasFees(request: UpdateGasFeesRequest) {
   const { txMeta } = request;
   const initialParams = { ...txMeta.txParams };
+  const advancedGasFees = !request.isAdvancedGasFeeDisabled
+    ? request.getAdvancedGasFee()
+    : undefined;
+
+  const shouldReadAdvancedGasFees =
```

Comment:
Rather than making this decision here and passing an additional boolean, I think the power of this architecture is that each `get` method makes the decisions based on the raw data.

If the goal is to avoid the duplicate swap type check, could we instead do that above and set `advancedGasFees` to `undefined` if it has a swap type.

Then all the `getX` methods have to do is do `if(savedGasFees)`?

Reply:
Done